File: problems/leetcode/2100/2101.cpp
// 2101. Detonate the Maximum Bombs
// https://leetcode.com/problems/detonate-the-maximum-bombs/

using namespace std;

#include <cmath>
#include <vector>

class Solution {
public:
  vector<vector<int>> connections;

  int maximumDetonation(vector<vector<int>> &bombs) {
    connections = vector<vector<int>>(bombs.size(), vector<int>());

    for (int i = 0; i < bombs.size(); i++) {
      int xi = bombs[i][0], yi = bombs[i][1], ri = bombs[i][2];

      for (int j = i + 1; j < bombs.size(); j++) {
        int xj = bombs[j][0], yj = bombs[j][1], rj = bombs[j][2];

        long long distanceSquare = pow(xj - xi, 2) + pow(yj - yi, 2);
        if (distanceSquare <= pow(ri, 2)) {
          connections[i].push_back(j);
        }
        if (distanceSquare <= pow(rj, 2)) {
          connections[j].push_back(i);
        }
      }
    }

    int maximumDetonationCount = 0;
    for (int i = 0; i < bombs.size(); i++) {
      vector<int> detonationCounts(bombs.size(), 0);
      vector<bool> visited(bombs.size(), false);
      maximumDetonationCount =
          max(maximumDetonationCount, dfs(detonationCounts, visited, i));
    }

    return maximumDetonationCount;
  }

  int dfs(vector<int> &detonationCounts, vector<bool> &visited, int index) {
    if (visited[index]) {
      return 0;
    }

    if (detonationCounts[index] != 0) {
      return detonationCounts[index];
    }

    visited[index] = true;
    int detonationCount = 1;
    for (auto nextBomb : connections[index]) {
      detonationCount += dfs(detonationCounts, visited, nextBomb);
    }

    detonationCounts[index] = detonationCount;
    return detonationCount;
  }
};

File: problems/leetcode/2100/2101/2101.py
# 2101. Detonate the Maximum Bombs
# https://leetcode.com/problems/detonate-the-maximum-bombs/description

from collections import deque
from typing import *

class Solution:
    def maximumDetonation(self, bombs: List[List[int]]) -> int:
        n = len(bombs)
        bombToNextBombs = dict()

        for i in range(n):
            bombToNextBombs[i] = [j for j in range(n) if self.isFirstBombAffectsSecondBomb(bombs[i], bombs[j])]

        maxBombs = 0
        for i in range(n):
            visited = set()
            visited.add(i)

            q = deque()
            q.append(i)

            while q:
                bomb = q.popleft()

                for nextBomb in bombToNextBombs[bomb]:
                    if nextBomb not in visited:
                        visited.add(nextBomb)
                        q.append(nextBomb)

            maxBombs = max(maxBombs, len(visited))

        return maxBombs

    def isFirstBombAffectsSecondBomb(self, bomb1: List[int], bomb2: List[int]) -> bool:
        x1, y1, r1 = bomb1
        x2, y2, _ = bomb2
        return (x1 - x2) ** 2 + (y1 - y2) ** 2 <= r1 ** 2
